[lldb] Don't create instance of `SymbolFileDWARFDebugMap` for non-Mach-O files

royitaqi opened the pull request:
# Change

`SymbolFileDWARFDebugMap::CreateInstance()` will return `nullptr` if the file is not a Mach-O.


# Benefit

This may improve **Linux** debugger launch time by skipping the creation of `SymbolFileDWARFDebugMap` during the [`SymbolFile::FindPlugin()` call](https://fburl.com/hi1w8dil), which loops through a list of `SymbolFile` plugins and tries to find the one that provides the best abilities.  If the `SymbolFileDWARFDebugMap` is created during this loop, it will load the symbol table of the file in question and loop through all the compile units in the debug map (the OSO entries) to calculate the abilities.


# Tests

Two unittests are added to verify the creation and the absence of `SymbolFileDWARFDebugMap` for Mach-O and non-Mach-O files, respectively.

Ran `ninja check-lldb` on both Darwin and Linux. No regression spotted. See results compared for [mac](https://pastebin.com/NdqTK4HC) and [linux](https://pastebin.com/bvjVJbi1).

```
$ ninja lldb-unit-test-deps
$ tools/lldb/unittests/SymbolFile/DWARF/SymbolFileDWARFTests --gtest_filter="*SymbolFileDWARFDebugMapTests*"
```

royitaqi committed:
[lldb] Do not create SymbolFileDWARFDebugMap for non-Mach-O files

royitaqi committed:
Add test, but failing

royitaqi committed:
Add/fix test

adrian-prantl commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARFDebugMap.cpp:
```diff
@@ -246,6 +247,8 @@ llvm::StringRef SymbolFileDWARFDebugMap::GetPluginDescriptionStatic() {
 
 SymbolFile *SymbolFileDWARFDebugMap::CreateInstance(ObjectFileSP objfile_sp) {
+  if (objfile_sp->GetPluginName() != ObjectFileMachO::GetPluginNameStatic())
+    return nullptr;
```
Who is calling this? I'm wondering if the check might make more sense at the call site?

JDevlieghere reviewed (changes requested):
Plugins shouldn't depend on each other. The correct way to do this would be to have a method in ObjectFile, something like `ObjectFile::SupportsDebugMap()` which returns false for all implementations except the `ObjectFileMachO` one.

bulbazord commented:
+1 to Jonas's comment. Plugins ideally don't depend on each other and they should strive to fail gracefully if they depend on a specific plugin instance that's not available. I recognize that many plugins depend on each other today, but I don't think we should be adding more direct dependencies if we can help it.

royitaqi commented:
@JDevlieghere  I see your point (some learning questions below about the principle).

However, I feel `ObjectFile::SupportsDebugMap()` is still closely tied to a specific plugin (not through compilation, but semantics). Imagine each plugin needs to add such a method into the `ObjectFile` class, then the class will be filled with plugin-specific methods like this, which I feel breaks the API scalability the plugin system wanted.

Instead, maybe it's better if we frame "DebugMap" as an `ObjectFileCapability` enum. Other capabilities can be added, too, e.g. does the object file imply split dwarf, etc.  Then the API will be `ObjectFile::GetSupportedCapabilities()`.

WDYT?

(Note: I am trying to think about how does this relate to the `CalculateAbilities()` function in those `SymbolFile` subclasses. I feel they are different, so it should be okay.)


-- (my learning question about the principle of "Plugins shouldn't depend on each other") --

I guess I'm missing a lot of context here. From my limited and naive view, there *are* natural connections between different kinds of plugins because how they usually work together in existing systems. E.g. macOS usually use Mach-O files and can use either `SymbolFileDWARF` and `SymbolFileDWARFDebugMap`, but never split-DWARF (dwp/dwo).   I don't understand the design principle of "Plugins shouldn't depend on each other". It feels like it's trying to ignore the natural connection between these plugins. I'm guessing we have that principle so as to avoid having to model a complex (and dynamic) dependency problem? (E.g. "A and B" and "B and C" should always work together, but "A and C" should never work together.)

labath commented:
FWIW, I completely agree @royitaqi. I'm not saying inter-plugin dependencies should be a free-for-all (*), but I think that some kinds of dependencies make sense and I definitely think this is one of them. The debug map format is intimately tied to the MachO and darwin ecosystems and I think there's approximately zero chance that it will be implemented elsewhere. Using the debug map requires MachO files, so why not just admit that?

This isn't the only example of that. For example, ProcessElfCore requires *elf* core files. ProcessMachCore requires MachO core files. And I think that's a perfectly natural state of the world and doing anything else would make things a lot more convoluted. I think something like `SupportsDebugMap` is a step in the wrong direction. It's true that removes the link time dependency but the moral dependency remains, and in fact this makes it worse, because now *every* object file plugin needs to know about the debug map. Why should ObjectFilePECOFF need to think about whether it supports a random symbol file plugin?

(*) I think that plugin dependencies should form a DAG. The fact that SymbolFileDWARFDebugMap depends on ObjectFileMachO means that ObjectFileMachO should not depend on SymbolFileDWARFDebugMap -- which I think is also perfectly natural restriction. And even if there is a reason why you might want to depend the other way, you can try to break it using the usual dependency reversal techniques, but you'd only need to touch these two plugins -- not every plugin of the two kinds.

I might also say that, for the sake of clarity, the dependency direction should be uniform across all plugins of the same type. So that SymbolFileDWARFDebugMap depends on ObjectFileMachO means that no other object file plugin can depend on any other symbol file plugin, which is, again, I think a reasonable requirement, and I would say that if we ever find a case where we have two plugin pairs which want to depend in the opposite directions, that this means that those plugin types are not designed properly (maybe they ought to be a single plugin?).

JDevlieghere commented:
> FWIW, I completely agree @royitaqi. I'm not saying inter-plugin dependencies should be a free-for-all (*), but I think that some kinds of dependencies make sense and I definitely think this is one of them. The debug map format is intimately tied to the MachO and darwin ecosystems and I think there's approximately zero chance that it will be implemented elsewhere. Using the debug map requires MachO files, so why not just admit that?                                                                        

To answer your last question: because the plugins (which is really a misnomer), as I understand it, are there to provide abstraction: the caller shouldn't have to know what concrete instance an `ObjectFile` is. The no-plugins-dependency enforces that at compile/link time. 

I'll be the first to admit that the current approach has shortcomings (like it does here, I totally agree with that). However we should also remember that it exists for a reason and that a lot of time and effort has been spent on detangling the plugins to fix layering issues and reduce binary sizes. The current rule is easy to understand and easy to enforce which makes it appealing and ensures we don't make things worse as we're still working on this. 

IMHO, the downsides being listed here don't outweigh the benefits, especially since we have a bunch of examples of those "moral dependencies". That said, if there's a way to enforce the DAG described in the asterisk, I'm totally fine with changing the rule to allow inter-plugin dependencies "where it makes sense" (which is still subjective). I made a similar argument in another PR a while ago, but I don't think we can expect reviewers to mentally verify this DAG every time someone tries to introduce a new inter-plugin dependencies.

TL;DR: My stance is that if there's a way to enforce the DAG, I'm totally fine with allowing inter-plugin dependencies where they make sense. Barring that I think the current rule, albeit overly strict, is the easiest way to enforce we don't regress inter-plugin dependencies in general. 

PS: I'm actually glad we're having this discussion. Whatever the outcome, I'd like to [document it on the website](https://lldb.llvm.org/resources/contributing.html) so contributors can know what to expect going forward. This comes up periodically and there's no reason this should be a surprise for existing or  new contributors.

labath commented:
> To answer your last question: because the plugins (which is really a misnomer), as I understand it, are there to provide abstraction: the caller shouldn't have to know what concrete instance an `ObjectFile` is. The no-plugins-dependency enforces that at compile/link time.

This is my problem with that. You're saying the "caller shouldn't have to *know*" -- which is a statement about his state of mind. We can't enforce the state of mind through syntactic restrictions. We can prevent the caller from *saying* he expects a specific subclass, but there's not way to prevent code from expecting that implicitly (perhaps even without it knowing). Functions like `SupportsDebugMap` make it worse because they encourage writing code like that and that makes the plugin system (the abstractions created by it) less useful as a whole. Like, if I now wanted to create a new symbol file plugin that works only with a specific object file, do I add `SupportsPsym` to every object file plugin? Or we don't have to look at new plugins, we have two good examples already: SymbolFileBreakpad and SymbolFileJSON. Both of these currently depend on the their respective object file plugin. Should that be replaced by something else (and would the result be better if we did)?

Nonetheless, I agree that plugins should create abstractions, but I think it's important to think about who are those abstractions directed at. The core code definitely shouldn't need to "look behind the curtain" (ideally not even through calls like `SupportsDebugMap`), but I don't think that means that noone else isn't allowed to do that. This situation is really simple, so it doesn't demonstrate the full scope of the problem, which is why I want to go back to the ProcessCore->ObjectFile example.

ProcessElfCore needs to extract a lot of detailed information from ObjectFileELF. Same goes for ProcessMachCore. At some level, the information these two pairs are exchanging is the same (threads, registers, ...), so one could imagine a generic interface to pass this -- but would that help anything? I say "no", because you still want to have the two plugins to work in pair -- if you didn't, then you wouldn't need separate process core plugins. This setup would make sense if we had a generic ProcessCore class (not a plugin) which works off of information provided by the object plugins -- but that's not the situation we're in now. You can make the same case for dynamic loaders -- should MacOSX-DYLD work with ObjectFilePECOFF?


> TL;DR: My stance is that if there's a way to enforce the DAG, I'm totally fine with allowing inter-plugin dependencies where they make sense. Barring that I think the current rule, albeit overly strict, is the easiest way to enforce we don't regress inter-plugin dependencies in general.

Enforcement is tricky, because "where it makes sense" is obviously a judgement call. It's also complicated by the fact that current graph is not a DAG. Linux linkers actually kind of enforce the graph implicitly, but since it's not a DAG now, we have a [workaround](https://github.com/llvm/llvm-project/blob/5c7bc6a0e6a5093812e6aa9719f7d98d14bb0015/lldb/source/Core/CMakeLists.txt#L92) and that prevents us catching any new bad deps. I guess it would be possible to do something similar to `NO_PLUGIN_DEPENDENCIES` in cmake, which only allows dependencies that we consider acceptable, though it's not foolproof: for example SymbolFileJSON doesn't list ObjectFileJSON in its cmake dependencies even though it definitely needs it.

I think the best we can do is some cmake check plus some documentation about which dependencies we consider "okay".
I don't think this can be foolproof though. Going by the examples above, I would say that we should permit dependencies on object file plugins from SymbolFiles, DynamicLoaders and Processes -- but there are still exceptions to that. For example, ProcessGdbRemote and DynamicLoaderStatic (and *maybe* SymbolFileDWARF) probably shouldn't depend on any specific ObjectFile plugin since they're meant to work with more than one (I guess that could be a general rule -- but it's not enforcable one).

A rule like "there shall be no inter-plugin dependencies" is easy to enforce and that makes it appealing, but I think it will be very hard for us to come into compliance with that (and I don't think we'd like the result even if we did). If we say some deps are okay, then a lot of things become automatically compliant, and we can focus on the high-value high-complexity deps like SymbolFile<->TypeSystem.

JDevlieghere commented:
> This is my problem with that. You're saying the "caller shouldn't have to _know_" -- which is a statement about his state of mind. We can't enforce the state of mind through syntactic restrictions. We can prevent the caller from _saying_ he expects a specific subclass, but there's not way to prevent code from expecting that implicitly (perhaps even without it knowing). Functions like `SupportsDebugMap` make it worse because they encourage writing code like that and that makes the plugin system (the abstractions created by it) less useful as a whole. Like, if I now wanted to create a new symbol file plugin that works only with a specific object file, do I add `SupportsPsym` to every object file plugin? Or we don't have to look at new plugins, we have two good examples already: SymbolFileBreakpad and SymbolFileJSON. Both of these currently depend on the their respective object file plugin. Should that be replaced by something else (and would the result be better if we did)?
> 
> Nonetheless, I agree that plugins should create abstractions, but I think it's important to think about who are those abstractions directed at. The core code definitely shouldn't need to "look behind the curtain" (ideally not even through calls like `SupportsDebugMap`), but I don't think that means that noone else isn't allowed to do that. This situation is really simple, so it doesn't demonstrate the full scope of the problem, which is why I want to go back to the ProcessCore->ObjectFile example.
> 
> ProcessElfCore needs to extract a lot of detailed information from ObjectFileELF. Same goes for ProcessMachCore. At some level, the information these two pairs are exchanging is the same (threads, registers, ...), so one could imagine a generic interface to pass this -- but would that help anything? I say "no", because you still want to have the two plugins to work in pair -- if you didn't, then you wouldn't need separate process core plugins. This setup would make sense if we had a generic ProcessCore class (not a plugin) which works off of information provided by the object plugins -- but that's not the situation we're in now. You can make the same case for dynamic loaders -- should MacOSX-DYLD work with ObjectFilePECOFF?

To be clear, I pretty much agree with everything you've said. I'm not arguing that the current approach is better, at least not on merit. My point is that barring a better alternative, those shortcoming are the price to pay to avoid the free-for-all nobody wants. I'm glad that you're offering an alternative below. 

> > TL;DR: My stance is that if there's a way to enforce the DAG, I'm totally fine with allowing inter-plugin dependencies where they make sense. Barring that I think the current rule, albeit overly strict, is the easiest way to enforce we don't regress inter-plugin dependencies in general.
> 
> Enforcement is tricky, because "where it makes sense" is obviously a judgement call. It's also complicated by the fact that current graph is not a DAG. Linux linkers actually kind of enforce the graph implicitly, but since it's not a DAG now, we have a [workaround](https://github.com/llvm/llvm-project/blob/5c7bc6a0e6a5093812e6aa9719f7d98d14bb0015/lldb/source/Core/CMakeLists.txt#L92) and that prevents us catching any new bad deps. I guess it would be possible to do something similar to `NO_PLUGIN_DEPENDENCIES` in cmake, which only allows dependencies that we consider acceptable, though it's not foolproof: for example SymbolFileJSON doesn't list ObjectFileJSON in its cmake dependencies even though it definitely needs it.
> 
> I think the best we can do is some cmake check plus some documentation about which dependencies we consider "okay". I don't think this can be foolproof though. Going by the examples above, I would say that we should permit dependencies on object file plugins from SymbolFiles, DynamicLoaders and Processes -- but there are still exceptions to that. For example, ProcessGdbRemote and DynamicLoaderStatic (and _maybe_ SymbolFileDWARF) probably shouldn't depend on any specific ObjectFile plugin since they're meant to work with more than one (I guess that could be a general rule -- but it's not enforcable one).

I'm fine with that. What I'm looking for is something that's (relatively) easy to understand and (relatively) easy to spot in code reviews. I'd love to have something foolproof (or work towards that) but I acknowledge that given the current state of things that's not achievable. What you describe here checks both boxes. 

> A rule like "there shall be no inter-plugin dependencies" is easy to enforce and that makes it appealing, but I think it will be very hard for us to come into compliance with that (and I don't think we'd like the result even if we did). If we say some deps are okay, then a lot of things become automatically compliant, and we can focus on the high-value high-complexity deps like SymbolFile<->TypeSystem.

Ack, agreed. 

To make this actionable, @labath or @royitaqi, is either of you willing to sign up to add the CMakery and document the new policy?

labath commented:
I'm glad we're in (abstract) agreement. As much as I would like to, I can't promise to do this by any particular date. I'd can *try*, but I don't know when will that happen.

In a way, I feel that this patch isn't the a good reason to motivate this (although I think that legalising the existing core file dependencies *is* a good reason) for this as I think there are acceptable workarounds (or even not workarounds) for the problem. In particular, I think that the solution from https://github.com/llvm/llvm-project/pull/139554 which looks at the Triple is okay. The Triple object even has a `isBinFormatMachO` method...

royitaqi committed:
Use objfile_sp->GetArchitecture().GetTriple().isAppleMachO() instead

royitaqi commented on lldb/unittests/SymbolFile/DWARF/SymbolFileDWARFDebugMapTests.cpp:
```diff
@@ -0,0 +1,154 @@
+  // The debug map should be non-null, because the file is Apple Mach-O.
+  auto debug_map =
+      SymbolFileDWARFDebugMap::CreateInstance(object_file->shared_from_this());
+  ASSERT_NE(debug_map, nullptr);
```
Currently this test fails because of this assertion. The reason is that the Triple in the object file is "arm64-apple-" (`ObjectFormat` being `ELF`).

I have verified:
1. The Triple in `module_spec` (line 76) is the expected "arm64-apple-macosx15.0.0".
2. But after creating the module (line 79), the Triple in the module is "arm64-apple-".

I wonder if this is because the file spec has no dir/filename and we are only providing data pointer.

Looking for advice from folks about how to make this test work.

royitaqi replied:
This is fixed in the new comment, by adding a load command `LC_BUILD_VERSION` into the mach-o yaml data.

royitaqi committed:
Update test

royitaqi commented:
Updated the code to using `Triple` of the `ObjectFile` (thanks to @labath for the suggestion). Updated the test.

@labath / @JDevlieghere: Please kindly take a look when you have the time.

JDevlieghere reviewed (approved):
LGTM. Thanks!

royitaqi committed:
Merge branch 'llvm:main' into dont-create-debug-map-on-linux

royitaqi committed:
Add assert and fix comment

royitaqi commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARFDebugMap.cpp:
```diff
@@ -246,6 +246,9 @@ llvm::StringRef SymbolFileDWARFDebugMap::GetPluginDescriptionStatic() {
 
 SymbolFile *SymbolFileDWARFDebugMap::CreateInstance(ObjectFileSP objfile_sp) {
+  // Don't create a debug map if the object file isn't a Mach-O.
+  if (!objfile_sp->GetArchitecture().GetTriple().isAppleMachO())
```
I think it makes sense to assert to make the requirement very clear. Added in latest commit. Thanks for the suggestion~!

--

Also fixed the comment in the test.

royitaqi commented:
> NB I reverted this on main because it depends on #142704 which I also reverted as it adds "-macho" to every Module's Triple string, and we need to find a way to mark the object file as MachO without doing that.

Hi @jasonmolenda,

I appreciate you actively trying to make the code base better, and I can see why the other revert may make sense, BUT:

> it depends on #142704

I respectfully disagree. **This patch doesn't depend on #142704.**
* In a normal production environment (e.g. on a regular macOS machine) and in tests, without #142704, the object format of mach-o files are already `MachO`.
* The only caveat of this patch may be if there are mach-o files out there which don't have the `LC_BUILD_VERSION` or `LC_VERSION_MIN_*` load commands. However, their triple is already wrong (object format being `ELF`), so their behavior is already undefined at best. So I don't think we need to worry about them.

Please kindly LMK if I missed something in the above.

Thanks,
Roy

jasonmolenda commented:
I'll admit I didn't debug it very far, but when I reverted the ObjectFileMachO.cpp change, `lldb/test/Shell/Breakpoint/case-sensitive.test` started failing on the bots.  When I built main myself and ran it, it failed on my desktop.  When I used the built lldb on the `case-sensitive.test.tmp` binary this test built, lldb could not set a source & line breakpoint -- it had no debug info.  When I also reverted this PR, the failure went away.

jasonmolenda commented:
e.g. https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/27150/ but it reproduced on my arm64 macOS desktop no problem when I did `ninja check-lldb-shell`

jasonmolenda commented:
FTR I suspect the correct fix for https://github.com/llvm/llvm-project/pull/142704 is

1. This Triple object file setting should be done in ArchSpec::SetArchitecture, so it doesn't depending on hitting an LC_BUILD load command, if anything is going to depend on it.
2. There needs to be a way of setting the Triple's object file to MachO (because it is defaulting to ELF when we start with just "arm64") without adding "-macho" to the triple string.  I don't see a way to do that with the llvm Triple methods today.

royitaqi commented:
@jasonmolenda I appreciate the details. IIRC I ran `ninja check-lldb` both before and after rebasing the patch onto #142704, but from what you said I must have not done that. Let me double check on my side.

royitaqi commented:
You are right, `lldb/test/Shell/Breakpoint/case-sensitive.test` failed on my arm64 macOS, too. So what I did must be that I tested before switching to using `isAppleMachO()`, then never tested it again until rebased on top of #142704.

Thank you for catching this and the revert. My bad (mistakenly thought the patch is independent).

royitaqi commented:
> the correct fix for #142704

How about:
1. In `Triple`'s `getDefaultFormat`, first check if vendor is `llvm::Triple::Apple` and return `MachO` if so. The vendor is set to `Apple` by `ObjectFileMachO::GetAllArchSpecs` calling `base_arch.SetArchitecture`, which calls `m_triple.setVendor(llvm::Triple::Apple)`.
